Add apphost customization

sbomer opened the pull request:
This adds logic to copy resources from the managed dll to the apphost.
- `_ComputeNETCoreBuildOutputFiles` now runs later in the build, at a point where `IntermediateAssembly` is available.
- The resource updating can only be done on windows hosts. When building for windows from a unix host, we log a warning but don't fail the build.
  - I introduced a string resource (`AppHostCustomizationRequiresWindowsHostWarning`) for this case, but not for errors specific to the `ResourceUpdater`
- No attempt is made to update resources when targeting unix, and no warning is shown in this case. Resources will just end up in the managed dll like before.
- I added a testcase that uses the `ResourceUpdater` to copy over and check a version resource.
- I did manual testing to verify that app icons and manifests work correctly with the change (the apphost icon shows up in the file explorer, and I was able to make an apphost ask for admin privileges on launch using an embedded app manifest).
- This also includes a method that @aaronrobinsonmsft should be able to use for adding a manifest for COM activation (see `AddResource`).

Note that this feature doesn't yet work with incremental build (see the discussion here https://github.com/dotnet/sdk/pull/2467).

Addresses https://github.com/dotnet/sdk/issues/1899.

@nguerrera @vitek-karas @jeffschwmsft

sbomer committed:
Add apphost customization

sbomer committed:
Pass intermediate assembly to apphost

sbomer committed:
Change task order to make IntermediateAssembly available to app host

sbomer committed:
Make intermediate assembly optional and fix version check test

sbomer committed:
Factor out PE image check

sbomer committed:
Check PE image and host to provide appropriate warning

This will emit a warning if host customization is requested but we are
building on non-Windows (where the required APIs don't exist).

sbomer committed:
Clean up resource updater a bit

sbomer committed:
Clean up ResourceUpdater

Add proper error handling and documentation

sbomer committed:
Run the resource test on windows only

AaronRobinsonMSFT commented on src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs:
```diff
@@ -0,0 +1,320 @@
+
+        private static void ThrowExceptionForInvalidUpdate()
+        {
+            throw new InvalidOperationException("Update handle is invalid. This instance may not be used for further updates.");
```
What is our localization story for this code base? Exception messages are normally localized.

sbomer replied:
I wasn't sure if it makes sense to take up NETSDK error codes for these exceptions since they should only fire if someone tries to use ResourceUpdater incorrectly. Some of the other non-localized exceptions are implementation details of the updater that could change if we make it possible to customize a windows apphost from unix. @nguerrera do you have a recommendation?

nguerrera replied:
We have a few error messages that are generally not possible without a bug, and we've just gone ahead and given them codes and let them be translated. I think it would be fine to do that here too.

---
In reply to: [219338566](https://github.com/dotnet/sdk/pull/2545#discussion_r219338566) [](ancestors = 219338566)

sbomer replied:
Went ahead and made them into localized strings.

AaronRobinsonMSFT commented on src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs:
```diff
@@ -0,0 +1,320 @@
+    public class ResourceUpdater
+    {
+        //
+        // Native methods for updating resources
```
Standard practice has been to put all P/Invokes and associated code in a private sealed class. It then allows use of the `nameof` operator and scopes functions properly. Unsure how important that is in this code base.

```
private sealed class Kernel32
{
        [DllImport(nameof(Kernel32), SetLastError=true)]
        private static extern IntPtr BeginUpdateResource(string pFileName,
                                                         [MarshalAs(UnmanagedType.Bool)]bool bDeleteExistingResources);
} 
```

nguerrera replied:
There have been a lot of different "standard" practices over the years. I'm happy to follow whatever is current in coreclr/corefx. We don't really have any interop in this repo yet apart from tests. I am deferring to you guys on how we should be doing it.

---
In reply to: [219338255](https://github.com/dotnet/sdk/pull/2545#discussion_r219338255) [](ancestors = 219338255)

nguerrera replied:
The class named after module + nameof is nice, but I don't feel strongly about it.

---
In reply to: [219643850](https://github.com/dotnet/sdk/pull/2545#discussion_r219643850) [](ancestors = 219643850,219338255)

sbomer replied:
I used your suggested convention.

vitek-karas commented on src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs:
```diff
@@ -0,0 +1,320 @@
+                ThrowExceptionForLastWin32Error();
+            }
+
+            // hResourceLoadedq is just a handle to the resource,
```
Nit: remove the trailing 'q' in `hResourceLoadedq`

vitek-karas commented on src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs:
```diff
@@ -0,0 +1,320 @@
+            return true;
+        }
+
+        private bool EnumLanguagesCallback(IntPtr hModule, IntPtr lpType, IntPtr lpName, ushort wLang, IntPtr lParam)
```
I would prefer more descriptive names for the Enum functions. They do enum, but they will eventually copy the resources as well.
Maybe something like:
EnumAndCopyToTarget_LanguagesCallback (NamesCallback, …).
With these names I simply expected it to build a list somewhere which the top caller would go over... which it doesn't.

vitek-karas commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToCustomizeTheApphost.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        [WindowsOnlyFact]
+        public void It_contains_resources_from_the_managed_dll()
```
I think we should test both file version and icon in some way - since those are our primary scenarios.
Maybe even the manifest...
I know that the test are essentially doing the same thing - but it's covering the end-to-end behaviors we want from this.

nguerrera replied:
I agree. It would be nice to have tests that demonstrate the end-to-end usage of all of these.

---
In reply to: [219612834](https://github.com/dotnet/sdk/pull/2545#discussion_r219612834) [](ancestors = 219612834)

sbomer replied:
I don't think this is worth doing due to the amount of work it would take to check that the icon or manifest is correct in the output - unless there's an existing end-to-end test for resources in roslyn, but I couldn't find anything like that.

nguerrera replied:
If it's too costly, I can live with not testing this.

vitek-karas replied:
Same here - if you tried and it looks like a lot of work - manual testing it is for now.

vitek-karas commented on src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs:
```diff
@@ -0,0 +1,320 @@
+    /// in a PE image. It currently only works on Windows, because it
+    /// requires various kernel32 APIs.
+    /// </summary>
+    public class ResourceUpdater
```
I think this should implement `IDisposable` since it owns the update handle. In case of exceptions this will likely leak that handle.

sbomer replied:
I made it IDisposable, using a new SafeUpdateHandle.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs:
```diff
@@ -0,0 +1,320 @@
+            IntPtr lpResourceData = LockResource(hResourceLoaded);
+            if (lpResourceData == IntPtr.Zero)
+            {
+                throw new Exception("Failed to lock resource.");
```
Please pick a more specific type than System.Exception

sbomer replied:
Introduced ResourceNotAvailableException for this.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/CreateAppHost.cs:
```diff
@@ -42,7 +44,9 @@ protected override void ExecuteCore()
                         WindowsGraphicalUserInterface = WindowsGraphicalUserInterface
-                    });
+                    },
+                    intermediateAssembly: IntermediateAssembly,
```
With this change, I'm wondering if we need to keep AppHostOptions. Seems that would have been more interesting if we were going to do the resources via that. Maybe we should just convert it to a bool?

sbomer replied:
I took your suggestion and removed AppHostOptions.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/CreateAppHost.cs:
```diff
@@ -21,6 +21,8 @@ public class CreateAppHost : TaskBase
         [Required]
         public string AppBinaryName { get; set; }
 
+        public string IntermediateAssembly { get; set; }
```
I think this should be [Required]

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -276,9 +276,8 @@ Copyright (c) .NET Foundation. All rights reserved.
-          DependsOnTargets="ResolvePackageAssets"
-          AfterTargets="ResolveReferences"
-          BeforeTargets="AssignTargetPaths"
+          DependsOnTargets="ResolvePackageAssets;CoreCompile"
```
>CoreCompile [](start = 49, length = 11)

I'm not sure about this dependency on CoreCompile. I know we need the intermediate assembly available, but there could be scenarios where we end up triggering a re-build when we were supposed to be using what is already on disk. This has come up a few times with other targets. I have to think about the best approach here.

nguerrera replied:
Still thinking about it, but I thing following the existing idioms in common targets would be to generate the apphost to intermediate directory upon/after compile and then designate the intermediate item as something to copy separately.

---
In reply to: [219644421](https://github.com/dotnet/sdk/pull/2545#discussion_r219644421) [](ancestors = 219644421)

sbomer replied:
@nguerrera generating the apphost into the intermediate directory would still have a dependency on CoreCompile, wouldn't it?

nguerrera replied:
Correct. 

I was thinking about changing the paradigm to react to compilation instead of depending on compilation. 

I am concerned that there may be places where this target is invoked without expecting compilation to happen even if the source files are out of date. Look at the long discussion on https://github.com/dotnet/cli/issues/5331 for some background on why this is worth worrying about.

What I'm wondering is: can the apphost generation happen in a separate target that is AfterTargets=CoreCompile (after we compile a managed exe to the intermediate directory, create the matching apphost next to it). And this target would just add the intermediate location to the set of files to be copied.

Adding @rainersigwald who may have other ideas here.

sbomer replied:
Agreed that if the apphost creation will be invoked in a way that doesn't expect compilation to happen, it shouldn't depend on CoreCompile. But the current target logically depends on an up-to-date IntermediateAssembly, and I think "DependsOnTargets=CoreCompile" captures that.

If the scenario you're thinking about is one where we do "publish --no-build", and expect that this will create an apphost for the previously-built intermediate assembly (even if source files are out of date), then I can see your concern. To me it makes sense to implement this by expressing the dependency on the outputs of a target that logically produces IntermediateAssembly, and having that target internally respect the --no-build option by using the existing IntermediateAssembly in that case. This way the --no-build behavior is transparent to targets that depend on build outputs.

AfterTargets=CoreCompile could work, but that has other issues. Any inputs to the apphost generator that don't come from CoreCompile should trigger apphost re-generation if they change. For example, setting the target to WinExe should cause the apphost to be re-generated with the gui bit, even if we don't want to re-compile.

nguerrera replied:
> If the scenario you're thinking about is one where we do "publish" --no-build

I'm thinking even more generally. This will cause any caller of GetCopyToOutputDirectoryItems to trigger compilation and I don't think that's OK. 

Agree that AfterTargets="CoreCompile" alone doesn't solve it. I'm still thinking about what's best here.




---
In reply to: [220379741](https://github.com/dotnet/sdk/pull/2545#discussion_r220379741) [](ancestors = 220379741)

sbomer replied:
It generally makes sense for GetCopyToOutputDirectoryItems to trigger compilation, because it logically depends on the compilation output through the apphost. If we want to optimize for "publish --no-build", we need to cut out some some dependency edges from Publish -> CoreCompile in this case (or alternatively make CoreCompile do nothing).

In https://github.com/dotnet/sdk/pull/2111, it looks like we already cut one edge from Publish -> CoreCompile by making Publish not depend on Build in the "--no-build" case. Maybe we want to do something similar here? We could make GetCopyToOutputDirectoryItems depend on _ComputeNETCoreBuildOutputFiles unless NoBuild is specified. If NoBuild is specified, GetCopyToOutputDirectoryItems could just try to get the last-built apphost from the intermediate directory.

Just an idea, but I'm sure there are more constraints I don't know about. I'll let you decide how you think this should be done - let me know if there's anything I can do to help!

rainersigwald replied:
`GetCopyToOutputDirectoryItems` should _not_ do actual work and should not depend on `CoreCompile` (and friends), because it is called on references in the `/p:BuildProjectReferences=false` case, which should not trigger compilation in the referenced projects. I don't think the `NoBuild` case applies here for that reason--you're building, but some other project.

Unfortunately, I can't think of a great way to accomplish this. Best I can think of is to move the actual work into a target (`CreateAppHost`?) and append `;CreateAppHost` to `$(CompileDependsOn)`, which would insert it in the normal flow shortly after `CoreCompile` but not require it to express the dependency. Then, do no real work in `_ComputeNETCoreBuildOutputFiles` and run it `BeforeTargets="GetCopyToOutputDirectoryItems"`.

nguerrera replied:
Thanks, @rainersigwald. This sounds like a good approach. @sbomer, can you look at doing it that way?

sbomer replied:
Thanks @rainersigwald and @nguerrera. I think the latest change does the trick - PTAL.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs:
```diff
@@ -0,0 +1,320 @@
+    /// in a PE image. It currently only works on Windows, because it
+    /// requires various kernel32 APIs.
+    /// </summary>
+    public class ResourceUpdater
```
The only classes that should be public in this assembly are the build tasks.

I missed in prior reviews that the AppHost class was made public (though you changed it to internal, which is good.)

Why is this being made public?

sbomer replied:
Just something I missed. I made it internal, thanks.

sbomer committed:
Use resource strings for all error messages

sbomer committed:
Use a private sealed class for pinvokes

sbomer committed:
Add comment about PE header

sbomer committed:
Document new parameters to AppHost.Create

sbomer committed:
Use more descriptive callback names

sbomer committed:
Implement IDisposable using a SafeHandle for hUpdate

sbomer committed:
Use custom exception type for LockResource failure

sbomer committed:
Remove AppHostOptions

sbomer committed:
Make IntermediateAssembly required in CreateAppHost task

sbomer committed:
Make ResourceUpdater internal

sbomer committed:
Don't throw managed exceptions from callbacks

Instead, use the extra parameter to capture error information, and
throw the appropriate error from the caller of EnumResourceTypes.

sbomer committed:
Move language-neutral constant to Kernel32 class

sbomer commented on src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs:
```diff
@@ -0,0 +1,320 @@
+            IntPtr hResource = FindResourceEx(hModule, lpType, lpName, wLang);
+            if (hResource == IntPtr.Zero)
+            {
+                ThrowExceptionForLastWin32Error();
```
Thank you. Fixed in https://github.com/dotnet/sdk/pull/2545/commits/ea4f506a0916f8523e32273ceff64f3ecd82a2a1. I also moved away from ThrowExceptionForHR due to https://github.com/dotnet/sdk/issues/2552.

nguerrera reviewed (changes requested):
Waiting for the issue with the CoreCompile dependency to be resolved.

sbomer committed:
Separate _CreateAppHost from _ComputeNETCoreBuildOutputFiles

This cuts the dependency from _ComputeNETCoreBuildOutputFiles on
CoreCompile. Instead, _CreateAppHost will run after compilation via
CompileDependson.

Locating the restored apphost has been factored into another target
_GetAppHostPaths, which is a dependency of both
_ComputeNETCoreBuildOutputFiles and _CreateAppHost. The destination
apphost path is also computed here, and passed to the CreateAppHost
task as an input.

This way, during a "publish --no-build", the apphost destintation
path (of the previously-generated apphost) is computed as a dependency
of _ComputeNETCoreBuildOutputFiles, and it is copied to the publish
directory without causing CoreCompile to run.

sbomer commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -267,16 +267,73 @@ Copyright (c) .NET Foundation. All rights reserved.
+    </ItemGroup>
+
+    <PropertyGroup Condition="'@(NativeRestoredAppHostNETCore)' != '' ">
+      <AppHostDestinationDirectoryPath>$(BaseIntermediateOutputPath)\$(TargetFramework)\$(RuntimeIdentifier)\host</AppHostDestinationDirectoryPath>
```
I didn't change the destination apphost path at all in this PR, but I can fix it with your suggestions. Why `$(IntermediateOutputPath)` instead of `$(BaseIntermediateOutputPath)`?
I don't know if there's a reason we put it in the host subfolder. Should I remove that?

nguerrera replied:
I see. I'm just reading this for the first time due to the diff. :)

We can change this separately.

sbomer replied:
Ok! I'm happy to do it here as well since it's small, just wanted to understand your suggestions.

nguerrera replied:
cc @peterhuene

peterhuene replied:
+1 on not needing the host subdirectory.  I can correct that later.

sbomer replied:
Thanks for the explanation. Changed to `IntermediateOutputPath`.

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -267,16 +267,73 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <!--
+    ============================================================
+                                        _CreateAppHost
+    If we found a restored apphost, create the modified destiniation apphost
```
typo: `destination`

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -267,16 +267,73 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <PropertyGroup>
+    <CompileDependsOn>
+      $(CompileDependsOn)
+      _CreateAppHost;
```
Sanity check: this file gets imported _after_ `Microsoft.Common.CurrentVersion.targets`, right?

nguerrera replied:
Right.

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -267,16 +267,73 @@ Copyright (c) .NET Foundation. All rights reserved.
+          DependsOnTargets="_GetAppHostPaths;CoreCompile"
+          Condition="'$(ComputeNETCoreBuildOutputFiles)' == 'true'">
+    <PropertyGroup>
+      <UseWindowsGraphicalUserInterface Condition="$(RuntimeIdentifier.StartsWith('win')) and '$(OutputType)'=='WinExe'">true</UseWindowsGraphicalUserInterface>
```
Does this need to be overrideable? Since it's set inside the target there's nothing the user can do to override it in their project, so perhaps it should have an underscore "private" name.

peterhuene replied:
This is a preexisting property from a previous commit, so I think we can make it "private" in a follow-up refactoring PR.

sbomer replied:
Made it private.

nguerrera commented:
I will investigate the tests failing on full msbuild leg. (At the moment, the new CI system does not publish test logs nor can it re-run individual legs, but I'm told this is getting resolved next week.)

nguerrera commented:
Test failure

<details>

```
Unit.net Console Runner (64-bit .NET Core 4.6.26515.07)
  Discovering: Microsoft.NET.Pack.Tests
  Discovered:  Microsoft.NET.Pack.Tests
  Starting:    Microsoft.NET.Pack.Tests
    Microsoft.NET.Pack.Tests.GivenThatWeWantToPackANetFrameworkLibrary.ImplicitReferencesAreNotIncludedAsFrameworkReferences [FAIL]
      Expected command to pass but it did not.
      File Name: C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\msbuild.exe
      Arguments: /t:Pack D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj
      Exit Code: 1
      StdOut:
      Microsoft (R) Build Engine version 15.9.14-preview+gae560092f5 for .NET Framework
      Copyright (C) Microsoft Corporation. All rights reserved.
      Build started 9/27/2018 11:50:44 AM.
      Project "D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj" on node 1 (Pack target(s)).
      _CheckForNETCoreSdkIsPreview:
      D:\Src\sdk\artifacts\Debug\bin\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(145,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj]
      PrepareForBuild:
        Creating directory "bin\Debug\net461\".
        Creating directory "obj\Debug\net461\".
      CoreCompile:
        C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Roslyn\csc.exe /noconfig /unsafe- /checked- /nowarn:1701,1702,1701,1702 /nostdlib+ /errorreport:prompt /warn:4 /define:TRACE;DEBUG;NETFRAMEWORK;NET461 /highentropyva+ /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\mscorlib.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.Core.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.Data.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.Drawing.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.IO.Compression.FileSystem.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.Numerics.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.Runtime.Serialization.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.Xml.dll" /reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.Xml.Linq.dll" /debug+ /debug:portable /filealign:512 /optimize- /out:obj\Debug\net461\PackImplicitReferences.dll /ruleset:"C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\Team Tools\Static Analysis Tools\\Rule Sets\MinimumRecommendedRules.ruleset" /subsystemversion:6.00 /target:library /warnaserror- /utf8output /deterministic+ PackImplicitReferences.cs "C:\Users\nicholg\AppData\Local\Temp\.NETFramework,Version=v4.6.1.AssemblyAttributes.cs" obj\Debug\net461\PackImplicitReferences.AssemblyInfo.cs /warnaserror+:NU1605
        Using shared compilation with compiler from directory: C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Roslyn
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(3261,7): error MSB4057: The target "_AfterCompileWinFXInternal [D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(3261,7): error MSB4057:      [D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(3261,7): error MSB4057:       _CreateAppHost" does not exist in the project. [D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj]
      Done Building Project "D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj" (Pack target(s)) -- FAILED.
      Build FAILED.
      "D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj" (Pack target) (1) ->
        C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(3261,7): error MSB4057: The target "_AfterCompileWinFXInternal [D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(3261,7): error MSB4057:      [D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(3261,7): error MSB4057:       _CreateAppHost" does not exist in the project. [D:\Src\sdk\artifacts\Debug\tmp\PackImplicitReferences\PackImplicitReferences\PackImplicitReferences.csproj]
          0 Warning(s)
          1 Error(s)
```

</details>

I think this has something to do with two-phase WPF builds. Investigating further.

nguerrera commented:
Oh, it looks it's just a missing semicolon here:

``` diff
    <CompileDependsOn>
-      $(CompileDependsOn)
+      $(CompileDependsOn);
      _CreateAppHost;
    </CompileDependsOn>

nguerrera commented:
After fixing that, there's another failure on my box. Investigating

<details>

```
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error APPX0002: Task 'ValidateConfiguration' failed. Value cannot be null. [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error APPX0002: Parameter name: input [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error APPX0002:  [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018: The "ValidateConfiguration" task failed unexpectedly. [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018: System.ArgumentNullException: Value cannot be null. [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018: Parameter name: input [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018:    at System.Version.Parse(String input) [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018:    at Microsoft.Build.AppxPackage.ValidateConfiguration.ExecuteImplementation() [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018:    at Microsoft.Build.AppxPackage.AppxPackagingTaskHelper.Execute(String file) [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018: --- End of stack trace from previous location where exception was thrown --- [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018:    at Microsoft.Build.AppxPackage.AppxPackagingTaskHelper.Execute(String file) [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018:    at Microsoft.Build.AppxPackage.ValidateConfiguration.Execute() [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
      C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\Microsoft\VisualStudio\v15.0\AppxPackage\Microsoft.AppXPackage.Targets(6068,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [D:\Src\sdk\artifacts\Debug\tmp\PCL_Project_r---890FB1DF\Dependency\Dependency.csproj]
```

</details>

nguerrera commented:
That second failure is internal bug 686365, which is fixed but my dogfood build hasn't updated. I believe the CI machine would not have a build of VS with that bug so this should be green once the semicolon is fixed.

sbomer committed:
Place customized apphost into IntermediateOutputPath

sbomer committed:
Make UseWindowsGraphicalUserInterface property private

And fix a typo

sbomer committed:
Add missing semicolon in CompileDependsOn

sbomer committed:
Improve warning message

sbomer committed:
Fix apphost incremental build

The apphost is (optionally) customized by copying resources from the
managed dll, and setting the GUI bit based on the Target
property (which also modifies the managed dll). Therefore, we use the
managed dll as an input for incremental build (the apphost needs to be
updated iff the managed dll was updated). This fixes
https://github.com/dotnet/sdk/issues/2554.

This also removes the unnecessary overwriteExisting flag from
AppHost.Create, fixing https://github.com/dotnet/sdk/issues/2473.

sbomer reviewed (commented):
@nguerrera thank you for investigating the CI failures!

sbomer commented on src/Tasks/Common/Resources/Strings.resx:
```diff
@@ -413,4 +413,20 @@ The following are names of parameters or literal values and should not be transl
-</root>
\ No newline at end of file
+  <data name="AppHostCustomizationRequiresWindowsHostWarning" xml:space="preserve">
+    <value>NETSDK1074: Unable to add resources to the application host executable because it's not being built on Windows.</value>
```
Good suggestion, fixed.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -341,6 +376,12 @@ Copyright (c) .NET Foundation. All rights reserved.
+    </AssignTargetPath>
+    <ItemGroup>
+      <_NoneWithTargetPath Include="@(_AllNETCoreCopyLocalItemsWithTargetPath)" />
+    </ItemGroup>
```
I don't follow yet. We're just computing paths here in this target now, not actually generating the apphost. Why can't that be done sooner?

rainersigwald replied:
No, I don't think there'd be a problem. Well-known/built-in metadata is evaluated just-in-time, so if you create an item, create the file, and then ask for `%(Whatever.ModifiedTime)` it should still work.

nguerrera replied:
I'm ok looking at this after the initial commit, btw. If this is green, we can merge as is.

sbomer replied:
I'll leave it as-is then. Thanks for all the help!

nguerrera replied:
Thank *you* for the contribution of this feature!

peterhuene commented:
It looks like some of the tests didn't get updated with the removal of `overwriteExisting`.

sbomer committed:
Remove overwriteExisting option from tests

sbomer commented:
Oops, left those changes out of my commit for some reason. Thanks.

nguerrera commented:
👏 🎉
